### PR TITLE
[SPARK-45092][SQL][UI] Avoid analyzing twice for failed queries

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

As a discussion starting from https://github.com/apache/spark/pull/42481#discussion_r1316776270, for failed queries, we need to avoid calling SparkPlanInfo fromSparkPlan, which triggers another round of analyzing.

This patch uses `Either[Throwable, () => T]` to pass the throwable conditionally and bypass plan explain functions on error.

### Why are the changes needed?

improvements of https://github.com/apache/spark/pull/42481

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/ui/UISeleniumSuite.scala [hunk 1]
```diff
@@ -110,7 +110,7 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser {
       val planDot = findAll(cssSelector(""".dot-file""")).map(_.text).toList
       assert(planDot.head.startsWith("digraph G {"))
       val planDetails = findAll(cssSelector("""#physical-plan-details""")).map(_.text).toList
-      assert(planDetails.head.contains("TABLE_OR_VIEW_NOT_FOUND"))
+      assert(planDetails.head.isEmpty)
```
Comment (yaooqinn) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/ui/UISeleniumSuite.scala:113]: Changes here according to https://github.com/apache/spark/pull/42481#discussion_r1316842538

**Comment by yaooqinn** [comment]
cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]
LGTM if all tests pass

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 2]
```diff
@@ -180,6 +189,18 @@ object SQLExecution extends Logging {
       sc.setLocalProperty(SPARK_JOB_INTERRUPT_ON_CANCEL, originalInterruptOnCancel)
     }
   }
+  def withNewExecutionId[T](
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:192]:
```suggestion

  def withNewExecutionId[T](
```

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**